File: 백준/알고리즘/브루트포스/N-Queen(9663)/solution.py
N = int(input())

answer = 0
col = [0] * N

def nQueen(row):
    global answer # 함수 내부에서 전역변수를 사용하기 위해서는 global 키워드를 사용하여 해당 전역 변수를 재선언해야함
    def promising(row):
        for i in range(row):
            if ( col[i] == col[row] or abs(i - row) == abs(col[i] - col[row])):
                return False
        return True
    
    if row == N:
        answer += 1
        return

    for i in range(N):
        col[row] = i

        if promising(row):
            nQueen(row + 1)

nQueen(0)
print(answer)

File: 백준/알고리즘/브루트포스/N-Queen(9663)/solution.js
const N = +require("fs").readFileSync("example.txt").toString().trim();
// const N = +require("fs").readFileSync("/dev/stdin").toString().trim()

let answer = 0;
let col = Array(N).fill(0);

function nQueen(row) {
  function promising(row) {
    for (let j = 0; j < row; j++) {
      if (
        col[j] == col[row] ||
        Math.abs(j - row) == Math.abs(col[j] - col[row])
      )
        return false;
    }

    return true;
  }

  if (row == N) {
    answer += 1;
    return;
  }

  for (let i = 0; i < N; i++) {
    col[row] = i;

    if (promising(row)) {
      nQueen(row + 1);
    }
  }
}

nQueen(0);
console.log(answer);
